src/java.base/share/classes/jdk/internal/module/SystemModuleFinders.java

```diff
@@ -253,6 +247,40 @@ private static ModuleFinder ofModuleInfos() {
         return new SystemModuleFinder(mrefs, nameToModule);
     }
 
+    /**
+     * Parses the {@code module-info.class} of all modules in the runtime image and
+     * returns a stream of {@link ModuleInfo.Attributes Attributes} for them. The
+     * returned attributes are in no specific order.
+     */
+    private static Stream<ModuleInfo.Attributes> allModuleAttributes() {
+        // System reader is a singleton and should not be closed by callers.
+        ImageReader reader = SystemImage.reader();
```

Comment:
The changes to SystemModuleFinders looks okay but I think we should drop the "System reader is a singleton and should not be closed by callers" here as it sets doubt that the ImageReader might be closed.

Reply:
Maybe, or the singleton returned by SystemImage.reader() is non-closeable. My only comment for now is tha the "should not be called by callers." is confusing here and would be better to remove. We do do this in the next change in the area if you want.